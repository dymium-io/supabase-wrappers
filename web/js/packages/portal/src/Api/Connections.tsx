import Alert from 'react-bootstrap/Alert'
import * as types from '@dymium/common/Types/Common'
import * as internal from '@dymium/common/Types/Internal'
import * as com from '../Common'

export function getConnections(setSpinner, setConns, setAlert, remap:internal.ConnectionMap|undefined, onSuccess) {
    setSpinner(true)
    setConns([])
    com.sendToServer("GET", "/api/getconnections",
        null, "",
        resp => {

            resp.json().then(_js => {
                let js = types.ConnectionResponse.fromJson(_js)
                if (js.status !== "OK") {
                    setAlert(
                        <Alert variant="danger" onClose={() => setAlert(<></>)} dismissible>
                            Error retrieving connections: {js.errormessage} { }
                        </Alert>
                    )
                    setTimeout(() => setSpinner(false), 500)
                    return
                }
            
                let cc = js.data.map(x => {
                    let ob = types.ConnectionRecord.fromJson({
                        id: x.id,
                        credid: x.credid,
                        dbtype: x.dbtype,
                        name: x.name,
                        dbname: x.dbname,
                        address: x.address,
                        port: x.port,
                        description: x.description,
                        useTLS: x.useTLS,

                    })
                    if(undefined != remap && x.name != null)
                        remap[x.name] = ob
                    return ob
                })

                setConns(cc)
                if(onSuccess != undefined) {
                    onSuccess()
                }
            })

            setTimeout(() => setSpinner(false), 500)
        },
        resp => {
            console.log("on error")
            setSpinner(false)
        },
        error => {
            console.log("on exception: " + error)
            setSpinner(false)
        })
}