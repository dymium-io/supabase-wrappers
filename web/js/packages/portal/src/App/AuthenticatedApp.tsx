import React, { useEffect } from 'react';
import { Outlet } from "react-router-dom";
import Row from 'react-bootstrap/Row'
import Col from 'react-bootstrap/Col'
import Backdrop from "../Backdrop"
import Menu from './Menu'
import Auth from '../Auth'
import { Link } from "react-router-dom";
import Sidebar from './Sidebar'
import * as com from '../Common'

const AuthenticatedApp = () => {
  let roles = com.getTokenProperty("roles")
  if (roles === null) {
    return <div className="py-0 my-0">
      <Backdrop />
      <div className="text-center" style={{
        position: 'absolute', top: '0px', left: '0px',
        width: '100%', height: '100vh'
      }}>
        <div id="loginbox" >
          <div className="row">
            <div className="col-sm ml-5  text-center">
              <h1 style={{ marginTop: '1.3em', fontSize: '5em', fontWeight: '300' }} className="logoheader ">
                <Link className="linklink" to="/" > Dymium </Link></h1>
            </div>
            <div className="col-sm mt-5  text-center">
              <h1 className="mt-5 logoheader">Error</h1>
              <h1 className="unauthorized" >Not authorized!</h1>
              <div className="pt-2 logofooter">
                You are succesfully authenticated but not authorized to use Dymium. Please contact your administrator.
              </div>

            </div>
          </div>

        </div>
      </div>



    </div>
  }
  return (
    <>
      <Auth />
      <Menu />

      <Row style={{ height: '100vh' }} className="p-0 m-0">
        <Col xs="auto" id="sidebar_container" className="px-0 sidebar_container"><Sidebar /></Col>
        <Col xs className="p-0 spider" style={{ overflowX: "scroll" }}>
          <Outlet />
        </Col>

      </Row>

    </>
  )
}

export default AuthenticatedApp;