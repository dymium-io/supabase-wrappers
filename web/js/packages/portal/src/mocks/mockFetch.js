import {queryconnection, querypolicy, querytable} from './queryconnection'

export default async function mockFetch(url, js) {
    console.log("in mock fetch: ", url)
    switch (url) {
        case "/api/getconnections": {
            return  new Promise( (resolve, reject) => resolve(
                {
                    ok: true,
                    status: 200,
                    json:  () => new Promise( (resolve, reject) => resolve(JSON.parse('{"status":"OK","errormessage":"","data":[{"name":"adventureworks","dbtype":"PostgreSQL","usesconnector":false,"connectorname":"","connectorid":"","tunnelname":"","tunnelid":"","address":"docker.for.mac.host.internal","port":5432,"dbname":"Adventureworks","useTLS":false,"description":"test data base from Microsoft","username":"postgres","password":null,"id":"ca7eb388-df2c-4b1f-a78e-35716061b63f","credid":"6d9e06fa-2660-44d6-b86a-83109b4812c3"},{"name":"northwind","dbtype":"PostgreSQL","usesconnector":false,"connectorname":"","connectorid":"","tunnelname":"","tunnelid":"","address":"docker.for.mac.host.internal","port":5432,"dbname":"northwind","useTLS":false,"description":"another MS database","username":"postgres","password":null,"id":"2e3f623c-64b7-428f-9186-9c4286936c8e","credid":"7e5b6636-8401-4579-a631-5ba6fa515861"},{"name":"aircrafts","dbtype":"PostgreSQL","usesconnector":false,"connectorname":"","connectorid":"","tunnelname":"","tunnelid":"","address":"docker.for.mac.host.internal","port":5432,"dbname":"aircrafts","useTLS":false,"description":"images","username":"dymium","password":null,"id":"64c59820-86fe-4ac0-b289-0170dce2519d","credid":"2f311469-db72-4f93-8bc4-468bdcaeff09"},{"name":"xxx","dbtype":"PostgreSQL","usesconnector":true,"connectorname":"Test","connectorid":"4a87fb02-f76c-4a0d-82b1-cfdcc23bdb6b","tunnelname":"test","tunnelid":"ba906750-0746-434a-acf0-d23c2d8f3fc6","address":"","port":0,"dbname":"adventureworks","useTLS":false,"description":"ccc","username":"dymium","password":null,"id":"45c9e82f-5d4b-41df-b798-b8e44fbcf680","credid":"9bff14c3-bccc-4c28-8ed5-3350fd22473b"},{"name":"aworks","dbtype":"PostgreSQL","usesconnector":false,"connectorname":"","connectorid":"","tunnelname":"","tunnelid":"","address":"docker.for.mac.host.internal","port":5432,"dbname":"Adventureworks","useTLS":false,"description":"MS test database, yes yes","username":"postgres","password":null,"id":"3f3f9866-77d2-41b4-8c7b-ef34c58fe97d","credid":"61e90024-2a15-402a-8991-d32de04c2a9b"},{"name":"111","dbtype":"PostgreSQL","usesconnector":false,"connectorname":"","connectorid":"","tunnelname":"","tunnelid":"","address":"docker.for.mac.host.internal","port":5432,"dbname":"Adventureworks","useTLS":false,"description":"xxx","username":"postgres","password":null,"id":"95da4e85-3e3e-46dc-8c1b-c8cc824f6d99","credid":"7d91f00c-6a14-40df-9834-b53cc3796883"},{"name":"sqlserver","dbtype":"SqlServer","usesconnector":false,"connectorname":"","connectorid":"","tunnelname":"","tunnelid":"","address":"host.docker.internal","port":1433,"dbname":"WideWorldImporters","useTLS":false,"description":"test sql server","username":"sa","password":null,"id":"80642caf-1319-41e7-9768-c06caab3ab45","credid":"650d9fcc-a96d-437b-9a7d-f78f8892182c"}]}') ))
                }
              )
            )
        }
        case "/api/getdatascopes" : {
            return  new Promise( (resolve, reject) => resolve(
                {
                    ok: true,
                    status: 200,
                    json:  () => new Promise( (resolve, reject) => resolve(JSON.parse('{"status":"OK","errormessage":"","records":[{"name":"test1","id":"50fe96d1-b1a4-48df-87bd-6aced2ab1f5a"},{"name":"test2","id":"b3fddf89-5822-4227-9948-3ecb3c8c2e33"}]}') ))
                }
              )
            )            
        }
        case "/api/getmappings" : {
            return  new Promise( (resolve, reject) => resolve(
                {
                    ok: true,
                    status: 200,
                    json:  () => new Promise( (resolve, reject) => resolve(JSON.parse('{"status":"OK","errormessage":"","records":[{"id":"788b3c71-5371-41e8-b135-8037bcfcfd9f","dymiumgroup":"foo","directorygroup":"bar","comments":"Admin group fubar"},{"id":"27713f32-6772-4509-9476-41099b11cb18","dymiumgroup":"dada","directorygroup":"production","comments":"Comment changed via test"}]}') ))
                }
              )
            )                     
        }
        
        case "/api/getgroupsfordatascopes" : {
            return  new Promise( (resolve, reject) => resolve(
                {
                    ok: true,
                    status: 200,
                    json:  () => new Promise( (resolve, reject) => resolve(JSON.parse('[{"id":"50fe96d1-b1a4-48df-87bd-6aced2ab1f5a","name":"test1","groupid":"788b3c71-5371-41e8-b135-8037bcfcfd9f","groupname":"foo"}]') ))
                }
              )
            )                     
        }        
        case "/api/updateconnection" : {
            expect(js).toEqual({
                method: 'POST',
                headers: { Authorization: 'Bearer mockJWT', Cache: 'no-cache' },
                body: '{"Id":"a2a4fce3-1db5-47ea-a084-4601f880b9ba","Name":"adventureworks","DbType":"PostgreSQL","Address":"docker.for.mac.host.internal","Port":5432,"Dbname":"Adventureworks","Description":"edited test data base from Microsoft"}'
              });
            return  new Promise( (resolve, reject) => resolve(
                {
                    ok: true,
                    status: 200,
                    json:  () => new Promise( (resolve, reject) => resolve(JSON.parse('{"status": "OK", "errormessage": "All is fine"}') ))
                }
              )
            )               
        }
        case "/api/deleteconnection": {
            expect(js).toEqual(
                {
                    method: 'POST',
                    headers: { Authorization: 'Bearer mockJWT', Cache: 'no-cache' },
                    body: '{"Id":"a2a4fce3-1db5-47ea-a084-4601f880b9ba"}'
                  }
            );
            return  new Promise( (resolve, reject) => resolve(
                {
                    ok: true,
                    status: 200,
                    json:  () => new Promise( (resolve, reject) => resolve(JSON.parse('{"status": "OK", "errormessage": ""}') ))
                }
              )
            )                
        }
        case "/api/queryconnection" : {
            expect(js).toEqual(
                {
                    method: 'POST',
                    headers: { Authorization: 'Bearer mockJWT', Cache: 'no-cache' },
                    body: '{"ConnectionId":"ca7eb388-df2c-4b1f-a78e-35716061b63f"}'
                  }
            );
            return  new Promise( (resolve, reject) => resolve(
                {
                    ok: true,
                    status: 200,
                    json:  () => new Promise( (resolve, reject) => resolve(JSON.parse(queryconnection) ))
                }
              )
            )    
            
        }
        case "/api/savedatascope" : {
            console.log(js)
            expect(js).toEqual(
                {
                    method: 'POST',
                    headers: { Authorization: 'Bearer mockJWT', Cache: 'no-cache' },
                    body: "{\"name\":\"testdb\",\"id\":null,\"records\":[{\"id\":null,\"connection\":\"adventureworks\",\"connectionId\":null,\"schema\":\"humanresources\",\"table\":\"employee\",\"typ\":\"integer\",\"position\":1,\"reference\":null,\"action\":\"Obfuscate\",\"col\":\"businessentityid\",\"semantics\":\"Business entity Id\",\"dflt\":null,\"isnullable\":false},{\"id\":null,\"connection\":\"adventureworks\",\"connectionId\":null,\"schema\":\"humanresources\",\"table\":\"employee\",\"typ\":\"varchar(15)\",\"position\":2,\"reference\":null,\"action\":\"Obfuscate\",\"col\":\"nationalidnumber\",\"semantics\":\"Social Security Number\",\"dflt\":null,\"isnullable\":false},{\"id\":null,\"connection\":\"adventureworks\",\"connectionId\":null,\"schema\":\"humanresources\",\"table\":\"employee\",\"typ\":\"varchar(256)\",\"position\":3,\"reference\":null,\"action\":\"Block\",\"col\":\"loginid\",\"semantics\":\"Login details\",\"dflt\":null,\"isnullable\":false},{\"id\":null,\"connection\":\"adventureworks\",\"connectionId\":null,\"schema\":\"humanresources\",\"table\":\"employee\",\"typ\":\"varchar(50)\",\"position\":6,\"reference\":null,\"action\":\"Obfuscate\",\"col\":\"jobtitle\",\"semantics\":\"Job position\",\"dflt\":null,\"isnullable\":false},{\"id\":null,\"connection\":\"adventureworks\",\"connectionId\":null,\"schema\":\"humanresources\",\"table\":\"employee\",\"typ\":\"date\",\"position\":7,\"reference\":null,\"action\":\"Allow\",\"col\":\"birthdate\",\"semantics\":\"N/A\",\"dflt\":null,\"isnullable\":false},{\"id\":null,\"connection\":\"adventureworks\",\"connectionId\":null,\"schema\":\"humanresources\",\"table\":\"employee\",\"typ\":\"character(1)\",\"position\":8,\"reference\":null,\"action\":\"Allow\",\"col\":\"maritalstatus\",\"semantics\":\"N/A\",\"dflt\":null,\"isnullable\":false},{\"id\":null,\"connection\":\"adventureworks\",\"connectionId\":null,\"schema\":\"humanresources\",\"table\":\"employee\",\"typ\":\"character(1)\",\"position\":9,\"reference\":null,\"action\":\"Obfuscate\",\"col\":\"gender\",\"semantics\":\"Gender\",\"dflt\":null,\"isnullable\":false},{\"id\":null,\"connection\":\"adventureworks\",\"connectionId\":null,\"schema\":\"humanresources\",\"table\":\"employee\",\"typ\":\"date\",\"position\":10,\"reference\":null,\"action\":\"Allow\",\"col\":\"hiredate\",\"semantics\":\"N/A\",\"dflt\":null,\"isnullable\":false},{\"id\":null,\"connection\":\"adventureworks\",\"connectionId\":null,\"schema\":\"humanresources\",\"table\":\"employee\",\"typ\":\"boolean\",\"position\":11,\"reference\":null,\"action\":\"Allow\",\"col\":\"salariedflag\",\"semantics\":\"N/A\",\"dflt\":\"true\",\"isnullable\":false},{\"id\":null,\"connection\":\"adventureworks\",\"connectionId\":null,\"schema\":\"humanresources\",\"table\":\"employee\",\"typ\":\"smallint\",\"position\":12,\"reference\":null,\"action\":\"Allow\",\"col\":\"vacationhours\",\"semantics\":\"N/A\",\"dflt\":\"0\",\"isnullable\":false},{\"id\":null,\"connection\":\"adventureworks\",\"connectionId\":null,\"schema\":\"humanresources\",\"table\":\"employee\",\"typ\":\"smallint\",\"position\":13,\"reference\":null,\"action\":\"Allow\",\"col\":\"sickleavehours\",\"semantics\":\"N/A\",\"dflt\":\"0\",\"isnullable\":false},{\"id\":null,\"connection\":\"adventureworks\",\"connectionId\":null,\"schema\":\"humanresources\",\"table\":\"employee\",\"typ\":\"boolean\",\"position\":14,\"reference\":null,\"action\":\"Allow\",\"col\":\"currentflag\",\"semantics\":\"N/A\",\"dflt\":\"true\",\"isnullable\":false},{\"id\":null,\"connection\":\"adventureworks\",\"connectionId\":null,\"schema\":\"humanresources\",\"table\":\"employee\",\"typ\":\"uuid\",\"position\":15,\"reference\":null,\"action\":\"Allow\",\"col\":\"rowguid\",\"semantics\":\"N/A\",\"dflt\":\"uuid_generate_v1()\",\"isnullable\":false},{\"id\":null,\"connection\":\"adventureworks\",\"connectionId\":null,\"schema\":\"humanresources\",\"table\":\"employee\",\"typ\":\"timestamp without time zone\",\"position\":16,\"reference\":null,\"action\":\"Allow\",\"col\":\"modifieddate\",\"semantics\":\"N/A\",\"dflt\":\"now()\",\"isnullable\":false},{\"id\":null,\"connection\":\"adventureworks\",\"connectionId\":null,\"schema\":\"humanresources\",\"table\":\"employee\",\"typ\":\"varchar\",\"position\":17,\"reference\":null,\"action\":\"Allow\",\"col\":\"organizationnode\",\"semantics\":\"N/A\",\"dflt\":\"'/'::character varying\",\"isnullable\":true}]}"
                  }
            );
            return  new Promise( (resolve, reject) => resolve(
                {
                    ok: true,
                    status: 200,
                    json:  () => new Promise( (resolve, reject) => resolve(JSON.parse(`{"status":"OK","errormessage":"Ghost Database created"}`) ))
                }
              )
            )    
            
        }        
        case "/api/querytable" : {
            console.log(js)
            expect(js).toEqual(
                {
                    method: 'POST',
                    headers: { Authorization: 'Bearer mockJWT', Cache: 'no-cache' },
                    body: '{"ConnectionId":"ca7eb388-df2c-4b1f-a78e-35716061b63f","Schema":"humanresources","Table":"shift"}'
                  }
            );
            return  new Promise( (resolve, reject) => resolve(
                {
                    ok: true,
                    status: 200,
                    json:  () => new Promise( (resolve, reject) => resolve(JSON.parse(querytable) ))
                }
              )
            )   
        }
        case "/api/getpolicies" : {
        
            return  new Promise( (resolve, reject) => resolve(
                {
                    ok: true,
                    status: 200,
                    json:  () => new Promise( (resolve, reject) => resolve(JSON.parse(querypolicy) ))
                }
              )
            )   
        }

        default: {
            throw new Error(`Unhandled request: ${url}`);
        }
    }
}

