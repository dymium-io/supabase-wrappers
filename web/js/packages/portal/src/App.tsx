import React, { useEffect } from 'react';
import LandingPage from './LandingPage'
import AuthenticatedApp from './App/AuthenticatedApp'
import Error404 from "./Error404"
import Logout from "./Logout"
import Error from "./Error"
import Invitation from "./Invitation"
import Sticky from "./App/Sticky"
import Dashboard from "./App/Dashboard"
import Customers from "./App/Customers"
import Connections from "./App/Connections"
import Datascopes from "./App/Datascopes"
import Groups from "./App/Groups"
import Rules from "./App/Rules"
import TestSQL from "./App/TestSQL"
import * as com from './Common'

import "@fontsource/roboto";
import 'bootstrap/dist/css/bootstrap.min.css';
import 'react-bootstrap-typeahead/css/Typeahead.css';

import { BrowserRouter, Routes, Route } from "react-router-dom";
//import Helmet from "react-helmet";

import '@dymium/common/App.scss';


function App() {
  return (
    <div className="App" id="app">

      <BrowserRouter>
        <Routes>

          <Route  path="/" element={<LandingPage />} />

          <Route path="/app/" element={<AuthenticatedApp />} >
            <Route path="/app/dashboard" element={<Dashboard />} />
            <Route path="/app/connections" element={<Connections />} />     
            <Route path="/app/datascopes" element={<Datascopes />} />                
            <Route path="/app/groups" element={<Groups />} />    
            <Route path="/app/rules" element={<Rules />} />         
            <Route path="/app/test" element={<TestSQL />} />                                             
            <Route path="/app/" element={<Sticky />} />          
          </Route>
          <Route  path="/app/logout" element={<Logout/>} />
          <Route path="/app/error" element={<Error />} />
          <Route path="/login" element={<Invitation />} />
          <Route path="*" element={<Error404 />} />
        </Routes>
      </BrowserRouter>

    </div>
  );
}

export default App;
