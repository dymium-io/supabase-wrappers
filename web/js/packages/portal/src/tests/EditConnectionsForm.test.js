/**
 * @jest-environment jsdom
 */
import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import {render, fireEvent, waitFor, screen } from '@testing-library/react'

import { BrowserRouter } from "react-router-dom";
import { act } from 'react-dom/test-utils';
import { Provider } from 'react-redux'
import store from '../Store/index'

import { EditConnections } from '../App/Connections'
import 'bootstrap/dist/css/bootstrap.min.css';
import '@dymium/common/App.scss';
import "@fortawesome/fontawesome-free/css/all.min.css";
import mockFetch from "../mocks/mockFetch";

beforeEach(() => {
// mock fetch
    global.fetch = jest.fn().mockImplementation(() => mockFetch)();
// create a fake JWT token
    global.sessionStorage.setItem("Session", "mockJWT");
})

afterEach(() => {
    jest.restoreAllMocks()
});

test('edit-connections-form', async () => {
    let component
    await act(async () => {
        component = render (
            <Provider store={store}>
                <BrowserRouter>
                    <EditConnections />
                </BrowserRouter>
            </Provider >
        )
    });

    // click on the edit button
    let find = component.getByRole("button", {name: "edit0"} )
    await act(async () => {    
        fireEvent.click(find)
    })
    // wait until Apply shows up
    await waitFor(() => {
        expect(screen.getByText('Apply')).toBeInTheDocument()
    }) 

    // take snapshot of the modal component
    let modal = component.getByTestId("modal-edit")
    expect(modal).toMatchSnapshot()
    
    // click on apply and test the update api call
    let apply = screen.getByText("Apply")
    await act(async () => {    
        fireEvent.click(apply)
    })
});
