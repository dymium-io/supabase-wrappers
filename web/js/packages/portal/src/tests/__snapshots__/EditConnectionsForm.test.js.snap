// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`delete-connections-form 1`] = `
<div
  class="modal-dialog modal-lg modal-dialog-centered"
  data-testid="modal-delete"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <div
        class="modal-title h4"
      >
        Delete connection 
        adventureworks
        ?
      </div>
      <button
        aria-label="Close"
        class="btn-close"
        type="button"
      />
    </div>
    <div
      class="modal-body"
    >
      Are you sure you want to remove the connection? This operation is irreversible.
    </div>
    <div
      class="modal-footer"
    >
      <button
        aria-label="Delete"
        class="btn btn-danger"
        data-testid="Delete"
        id="Delete"
        role="button"
        type="button"
      >
        Delete
      </button>
       
      <button
        class="btn btn-dymium"
        type="button"
      >
        Cancel
      </button>
    </div>
  </div>
</div>
`;

exports[`edit-connections-form 1`] = `
<div
  class="modal-dialog modal-lg"
  data-testid="modal-edit"
>
  <div
    class="modal-content"
  >
    <form
      class=""
      novalidate=""
    >
      <div
        class="modal-header"
      >
        <div
          class="modal-title h4"
        >
          Edit Data Source 
          adventureworks
        </div>
        <button
          aria-label="Close"
          class="btn-close"
          type="button"
        />
      </div>
      <div
        class="modal-body"
      >
        <div
          class="fade alert alert-danger alert-dismissible show"
          role="alert"
        >
          <button
            aria-label="Close alert"
            class="btn-close"
            type="button"
          />
          Error retrieving access key: 
          Unhandled request: /api/getconnectors
           
        </div>
        <div
          class="row"
        >
          <div
            class="col-auto"
          >
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="dbtype"
              >
                Data Source type:
              </label>
              <select
                class="form-control form-control-sm"
                id="dbtype"
                required=""
              >
                <option
                  value=""
                >
                  ...
                </option>
                <option
                  value="PostgreSQL"
                >
                  PostgreSQL Database
                </option>
                <option
                  value="MySQL"
                >
                  MySQL Database
                </option>
                <option
                  value="MariaDB"
                >
                  MariaDB Database
                </option>
                <option
                  value="SqlServer"
                >
                  Microsoft SQL Server
                </option>
                <option
                  value="OracleDB"
                >
                  Oracle Database
                </option>
                <option
                  value="DB2"
                >
                  IBM DB2 Database
                </option>
                <option
                  value="MongoDB"
                >
                  MongoDB Database
                </option>
                <option
                  value="Elasticsearch"
                >
                  Elasticsearch Database
                </option>
              </select>
              <div
                class="valid-feedback"
              >
                Looks good!
              </div>
              <div
                class="invalid-feedback"
              >
                Select Data Source type
              </div>
            </div>
          </div>
          <div
            class="col-auto"
          >
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="dname"
              >
                <div
                  class=""
                >
                  <div
                    class=" text-wrap"
                  >
                    Dymium
                     
                    <span
                      class="text-nowrap"
                    >
                      name
                      <i
                        class="ml-1 blue fas fa-question-circle text-newinfo"
                      />
                      :
                    </span>
                  </div>
                </div>
              </label>
              <input
                class="form-control form-control-sm"
                id="dname"
                pattern="[a-z0-9_]+"
                placeholder="alpha_num, small caps"
                required=""
                type="text"
                value="adventureworks"
              />
              <div
                class="valid-feedback"
              >
                Looks good!
              </div>
              <div
                class="invalid-feedback"
              >
                Type systemwide unique name to use in SQL
              </div>
            </div>
          </div>
          <div
            class="col-auto"
          >
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="dbname"
              >
                <div
                  class=""
                >
                  <div
                    class=" text-wrap"
                  >
                    Database
                     
                    <span
                      class="text-nowrap"
                    >
                      name
                      <i
                        class="ml-1 blue fas fa-question-circle text-newinfo"
                      />
                      :
                    </span>
                  </div>
                </div>
              </label>
              <input
                class="form-control form-control-sm"
                id="dbname"
                pattern=".+"
                placeholder="Alpha_Num"
                required=""
                type="text"
                value="Adventureworks"
              />
              <div
                class="valid-feedback"
              >
                Looks good!
              </div>
              <div
                class="invalid-feedback"
              >
                Type systemwide unique name to use in SQL
              </div>
            </div>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="mt-1 col"
            style="padding-left: 1.5em;"
          >
            <div
              class="react-toggle yellowtoggle"
            >
              <div
                class="react-toggle-track"
              >
                <div
                  class="react-toggle-track-check"
                >
                  <svg
                    height="11"
                    viewBox="0 0 14 11"
                    width="14"
                  >
                    <path
                      d="M11.264 0L5.26 6.004 2.103 2.847 0 4.95l5.26 5.26 8.108-8.107L11.264 0"
                      fill="#fff"
                      fill-rule="evenodd"
                    />
                  </svg>
                </div>
                <div
                  class="react-toggle-track-x"
                >
                  <svg
                    height="10"
                    viewBox="0 0 10 10"
                    width="10"
                  >
                    <path
                      d="M9.9 2.12L7.78 0 4.95 2.828 2.12 0 0 2.12l2.83 2.83L0 7.776 2.123 9.9 4.95 7.07 7.78 9.9 9.9 7.776 7.072 4.95 9.9 2.12"
                      fill="#fff"
                      fill-rule="evenodd"
                    />
                  </svg>
                </div>
              </div>
              <div
                class="react-toggle-thumb"
              />
              <input
                class="react-toggle-screenreader-only"
                id="conn-status"
                type="checkbox"
              />
            </div>
            <label
              class="form-check-label"
              for="conn-status"
              style="margin-left: 0.5em; position: relative; top: -0.38em;"
            >
              Use Connector instead of direct addressing
            </label>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="col-auto"
          >
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="ipaddress"
              >
                Address:
              </label>
              <input
                class="form-control form-control-sm"
                id="ipaddress"
                pattern="^[a-zA-Z0-9._-]+$"
                placeholder="DB IP address or host name"
                required=""
                type="text"
                value="docker.for.mac.host.internal"
              />
              <div
                class="valid-feedback"
              >
                Looks good!
              </div>
              <div
                class="invalid-feedback"
              >
                Enter DB address for Dymium
              </div>
            </div>
          </div>
          <div
            class="col-auto"
          >
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="portnumber"
              >
                Port:
              </label>
              <input
                class="form-control form-control-sm"
                id="portnumber"
                min="1"
                placeholder="DB port number"
                required=""
                type="number"
                value="5432"
              />
              <div
                class="valid-feedback"
              >
                Looks good!
              </div>
              <div
                class="invalid-feedback"
              >
                Select DB port for Dymium
              </div>
            </div>
          </div>
          <div
            class="col-auto"
            style="display: flex; align-items: bottom;"
          >
            <div
              style="margin-top: 1.8em;"
            >
              <div
                class="react-toggle yellowtoggle"
              >
                <div
                  class="react-toggle-track"
                >
                  <div
                    class="react-toggle-track-check"
                  >
                    <svg
                      height="11"
                      viewBox="0 0 14 11"
                      width="14"
                    >
                      <path
                        d="M11.264 0L5.26 6.004 2.103 2.847 0 4.95l5.26 5.26 8.108-8.107L11.264 0"
                        fill="#fff"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </div>
                  <div
                    class="react-toggle-track-x"
                  >
                    <svg
                      height="10"
                      viewBox="0 0 10 10"
                      width="10"
                    >
                      <path
                        d="M9.9 2.12L7.78 0 4.95 2.828 2.12 0 0 2.12l2.83 2.83L0 7.776 2.123 9.9 4.95 7.07 7.78 9.9 9.9 7.776 7.072 4.95 9.9 2.12"
                        fill="#fff"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </div>
                </div>
                <div
                  class="react-toggle-thumb"
                />
                <input
                  class="react-toggle-screenreader-only"
                  id="tls-status"
                  type="checkbox"
                />
              </div>
              <label
                class="form-check-label"
                for="tls-status"
                style="margin-left: 0.5em; position: relative; top: -0.38em;"
              >
                Use TLS
              </label>
            </div>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="col"
          >
            <div
              class="mb-1"
            >
              <div
                class="form-check"
                style="margin-top: 0.2em;"
              >
                <input
                  class="form-check-input"
                  id="changecred"
                  type="checkbox"
                />
                <label
                  class="form-check-label"
                  for="changecred"
                  title=""
                >
                  Change Data Source credentials
                </label>
              </div>
            </div>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="col"
          >
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="description"
              >
                Description:
              </label>
              <textarea
                class="form-control"
                id="description"
                placeholder="Please put in the description of this connection"
                required=""
                rows="3"
                style="width: 100%;"
              >
                test data base from Microsoft
              </textarea>
              <div
                class="valid-feedback"
              >
                Looks good!
              </div>
              <div
                class="invalid-feedback"
              >
                Please put in some description
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="modal-footer"
      >
        <button
          aria-label="Apply"
          class="btn btn-dymium"
          id="Apply"
          role="button"
          type="submit"
        >
          Apply
        </button>
         
        <button
          class="btn btn-dymium"
          type="button"
        >
          Cancel
        </button>
      </div>
    </form>
  </div>
</div>
`;
